restful_api/controllers/main.py

```diff
@@ -0,0 +1,262 @@
+        return o.isoformat()
+
+def validate_token(func):
+    """."""
+
+    @functools.wraps(func)
+    def wrap(self, *args, **kwargs):
+        """."""
+        access_token = request.httprequest.headers.get("Authorization")
+        if not access_token:
+            return invalid_response(
+                401, "missing http request (or) access token",)
+        
+        access_token_data = (
+                                request.env["api.access_token"]
+                                .sudo()
+                                .search([("token", "=", access_token)], order="id DESC", limit=1)
+                            )
+
+        if (
+                access_token_data.find_one_or_create_token(user_id=access_token_data.user_id.id)
+                != access_token
+            ):
+            return invalid_response("401", "token seems to have expired or invalid")
+
+        request.session.uid = access_token_data.user_id.id
+        request.uid = access_token_data.user_id.id
+        return func(self, *args, **kwargs)
+
+    return wrap
+
+_routes = ["/api/<model>", "/api/<model>/<id>", "/api/<model>/<id>/<action>"]
+            
+#-------------------------------- example of api methods ----------------------------------------#
+
+class APIController(http.Controller):
+    """."""
+
+    def __init__(self):
+        self._model = "ir.model"
+
+    @validate_token
+    @http.route(_routes, type="http", auth="none", methods=["GET","OPTIONS"], csrf=False, cors='*')
+    
+    def get(self, model=None, id=None, **payload):
+        ioc_name = model
+        model = request.env[self._model].sudo().search([("model", "=", model)], limit=1)
+        if model:
+            domain, fields, offset, limit, order = extract_arguments(payload)
+            data = (
+                request.env[model.model]
+                .sudo()
+                .search_read(
+                    domain=domain,
+                    fields=fields,
+                    offset=offset,
+                    limit=limit,
+                    order=order,
+                )
+            )
+            if id:
+                domain = [("id", "=", int(id))]
+                data = (
+                    request.env[model.model]
+                    .sudo()
+                    .search_read(
+                        domain=domain,
+                        fields=fields,
+                        offset=offset,
+                        limit=limit,
+                        order=order,
+                    )
+                )
+            if data:
+                return werkzeug.wrappers.Response(
+                status=200,
+                content_type="application/json; charset=utf-8",
+                headers=[("Cache-Control", "no-store"),('Access-Control-Allow-Headers', 'Origin, Content-Type, X-Auth-Token, charset'),('Access-Control-Allow-Methods','POST, GET, OPTIONS, DELETE, PUT'), ('Access-Control-Allow-Origin', 'http://localhost:8069'), ("Pragma", "no-cache")],
```

Comment:
Localhost 8069 is hardcoded here? Will this cause issues in non-localhost deployments?

Reply:
For development purposes, it was initially specified as 'localhost,' but this part has now been removed.